infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationDiscoveryService.java

```diff
@@ -436,40 +445,51 @@ public void processClusterRoleChange(TopologyDescriptor newTopology) {
      *
      * @param event discovery event
      */
-    private void processTopologyChangeNotification(DiscoveryServiceEvent event) {
+    public void processTopologyChangeNotification(DiscoveryServiceEvent event) {
         // Stale notification, skip
-        if (event.getTopologyConfig().getTopologyConfigID() < getReplicationManager().getTopology().getTopologyConfigId()) {
-            log.debug("Stale Topology Change Notification, current={}, received={}", topologyDescriptor.getTopologyConfigId(),
-                    event.getTopologyConfig());
+        if (event.getTopologyConfig().getTopologyConfigID() < topologyDescriptor.getTopologyConfigId()) {
+            log.debug("Stale Topology Change Notification, current={}, received={}",
+                    topologyDescriptor.getTopologyConfigId(), event.getTopologyConfig());
             return;
         }
 
         TopologyDescriptor newConfig = new TopologyDescriptor(event.getTopologyConfig());
-        topologyDescriptor = newConfig;
 
-        //On topology change notification, store latest topology and only process if current node is the leader
-        if (isLeader) {
-            if (newConfig.getTopologyConfigId() == getReplicationManager().getTopology().getTopologyConfigId()) {
-                if (localNodeDescriptor.getRoleType() == ClusterRole.STANDBY) {
-                    return;
-                }
+        // Note: should not update topology here. Otherwise, below check is always true.
+        if (newConfig.getTopologyConfigId() == topologyDescriptor.getTopologyConfigId()) {
+            if (localClusterDescriptor.getRole() == ClusterRole.STANDBY) {
+                return;
+            }
 
-                // If the current node is active, compare with the current topologyConfig, see if there are additional or
-                // removed standbys
-                getReplicationManager().processStandbyChange(newConfig);
-            } else {
-                // TODO: Are we sure that when there is a topologyConfigId change it implies a role change
-                //  and not a new standby added??
-                processClusterRoleChange(newConfig);
+            // If the current node is active, compare with the current topologyConfig, see if there are additional or
+            // removed standbys
+            if (replicationManager != null && isLeader) {
+                replicationManager.processStandbyChange(newConfig);
```

Comment:
I see the point that only the leader should do replication related jobs.
But discovery service does not only have topologyConfig, it also has cluster, node... etc.
So we need to make sure all fields get updated, and once any node gets a lock, it has updated configs...

Reply:
Good catch! This bug is introduced by Anny's new change. The new topology information should be updated on all the nodes, not just the leader.

Reply:
Correct, so one question. When a DISCOVERED_TOPOLOGY comes in (change in the topology) we come in this method processTopologyChangeNotification, only if it is stale we can ignore that incoming topology, otherwise, shouldn't the first thing we do is call your updateLocalTopology() to update all local objects? and maybe you can store the old topology in a temp for the further comparisons.

Reply:
Yeah, it is cleaner. Besides call updateLocalTopology(), I also need to update other components' config. Like logReplicationServerHandler.setActive

Reply:
@annym  @xiaoqin2012  I realized that updateLocalTopology() is dependent on config id change.

1. same config id -> add/remove standby, no role change. Process standby first then update configs

2. higher config id -> might have a role change. Need to stop all replications, which will use current status(role, config)
So we can not pdateLocalTopology() until we stop all replications... Besides, notify the metadata manager that we have a higher config id, update sink manager, and update server's role.